modules/starfish/2.0/starfish2.smk

```diff
@@ -0,0 +1,247 @@
+        )
+    return vcfs
+
+# Create a union VCF
+rule _starfish_union: 
+    input: 
+        _starfish_get_output
+    output: 
+        union = CFG["dirs"]["starfish"] + "{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}/{union_vcf}.vcf.gz", 
+        union_tbi = CFG["dirs"]["starfish"] + "{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}/{union_vcf}.vcf.gz.tbi"
+    log:
+        stdout = CFG["logs"]["starfish"] + "{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}/{union_vcf}.stdout.log",
+        stderr = CFG["logs"]["starfish"] + "{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}/{union_vcf}.stderr.log"
+    params: 
+        options = CFG["options"]["starfish_union"]
+    conda: 
+        CFG["conda_envs"]["bcftools"]
+    threads:
+        CFG["threads"]["starfish_union"]
+    resources:
+        **CFG["resources"]["starfish_union"]
+    shell: 
+        op.as_one_line("""
+        bcftools merge {params.options} --force-samples 
+        -Oz -o {output.union} 
+        {input} && 
+        tabix -p vcf {output.union}
+        """)
+
+# Symlinks the final output files into the module results directory (under '99-outputs/')
+rule _starfish_output_vcf:
+    input:
+        vcf = CFG['dirs']['starfish'] + "{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}/{vcf_name}.vcf.gz"
+    output:
+        vcf = CFG["dirs"]["outputs"] + "vcf/{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}.{vcf_name}.vcf.gz", 
+        tbi = CFG["dirs"]["outputs"] + "vcf/{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}.{vcf_name}.vcf.gz.tbi"
+    wildcard_constraints: 
+        vcf_name = "\S+(?<!union)"
+    run:
+        op.relative_symlink(input.vcf, output.vcf), 
+        op.relative_symlink(input.vcf + ".tbi", output.tbi)
+
+rule _starfish_output_union:
+    input:
+        vcf = CFG['dirs']['starfish'] + "{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}/{union_vcf}.vcf.gz", 
+        tbi = CFG['dirs']['starfish'] + "{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}/{union_vcf}.vcf.gz.tbi",
+    output:
+        vcf = CFG["dirs"]["outputs"] + "vcf/{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}.{union_vcf}.vcf.gz", 
+        tbi = CFG["dirs"]["outputs"] + "vcf/{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}.{union_vcf}.vcf.gz.tbi"
+    wildcard_constraints: 
+        union_vcf = union_vcf
+    run:
+        op.relative_symlink(input.vcf, output.vcf), 
+        op.relative_symlink(input.tbi, output.tbi)
+
+rule _starfish_output_venn: 
+    input: 
+        venn = str(rules._starfish_run.output.venn)
+    output: 
+        venn = CFG["dirs"]["outputs"] + "venn/{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}.venn.pdf"
+    run: 
+        op.relative_symlink(input.venn, output.venn)
+
```

Comment:
Can you please add one or two comment lines explaining what these complex functions do?

Reply:
To clarify... "get the checkpoint output directory" means literally nothing to me

Reply:
Thank you for adding detailed in-line comments in the code. It is much more clear now.